Compatibility changes for upcoming flask version 2.3

jrast opened the pull request:
This PR addresses the issues mentioned in https://github.com/vimalloc/flask-jwt-extended/issues/492

vimalloc commented:
Thanks for looking into this!

 Another option that might be worth considering is to follow suite with flask and depreciate the json encoder options as part of this extension, and mark that release as requiring flask >= 2.3. But that's just a thought I had when I woke up and without any caffeine so I want to think on it some more 😄

jrast commented:
Actually many other extensions will also run into problems: Flask-SQLALchemy (https://github.com/pallets-eco/flask-sqlalchemy/issues/1068), Flask-Security / Flask-Security-Too (https://github.com/pallets/flask/discussions/4711) just to name the few I use most. Flask 2.3 will require quite a bit of changes in extensions as so many things will be removed / restructured / change. Maybe it's worth to keep an eye on other extensions and see how they handle the issues.

tgross35 reviewed (commented):
Added notes on some little canonicalization things

jrast commented on flask_jwt_extended/utils.py:
```diff
@@ -23,7 +23,7 @@ def get_jwt() -> dict:
         The payload (claims) of the JWT in the current request
     """
-    decoded_jwt = getattr(_request_ctx_stack.top, "jwt", None)
+    decoded_jwt = getattr(g, "_jwt_extended_jwt", None)
```
I tried to stay as close to the original code as possible. @vimalloc should I change this to the cleaner proposed solution?

tgross35 replied:
Oh hey, I didn't realize you were the same person as on `flask_sqlalchemy`. I'll just link my to my reasoning there https://github.com/pallets-eco/flask-sqlalchemy/pull/1073#discussion_r944569555

vimalloc replied:
I think `.get` would be better as well 🎉

jrast commented on flask_jwt_extended/config.py:
```diff
@@ -9,12 +11,37 @@
 
 from flask import current_app
-from flask.json import JSONEncoder
+from flask.json.provider import _default
```
Agreed. Will change this.

vimalloc commented on flask_jwt_extended/config.py:
```diff
@@ -9,12 +11,37 @@
+        overriding how basic types like ``str`` or ``list`` are
+        serialized, they are handled before this method.
+        """
+        return _default(o)
```
Hrm... the thing I am sold on is that the only reason we have this `config.json_encoder` property is so that we can use the same encoder for the flask app as for pyjwt (per a8605502).

Even though this `_default` version does more then standard `json.JSONEncoder`, we could still run into situations where someone is using a custom thing that has additional features, and breaks when trying to use those custom things with a JWT.

I want to dig into the internals here a bit more and see if there is a way to bridge this gap. But if that isn't going to be easy or I don't have enough time to look, I'm happy to go with this approach initially as a stop gap solution!

👍

jrast replied:
The solution using `_default` is not optimal, as you said it might break some setups.

jrast replied:
Actually in Flask >= 2.2 there is the `json_provider_class` property on the `current_app`. I think this would be the way to go to support other JSON Providers. We just need to find a way to make this Flask <2.2 / >= 2.2 compatible... If this is even a goal?

vimalloc replied:
Would something as simple as having an `except AttributeError:` and using the old `current_app.json_encoder` in that case to support both versions? If so I would just as soon do that, but I'm not opposed to updating the `setup.py` to have the newer release of flask be the minimum supported version if it isn't so simple.

jrast replied:
Sadly it's not as simple as that. The base flask JSON provider class (https://github.com/pallets/flask/blob/main/src/flask/json/provider.py#L20) does not require that custom implementations provide a `default` method, so it cannot be expected that this method exists.

vimalloc reviewed (commented):
Sorry this took me so long to look at, it's been a crazy week over here.

jrast commented:
I'm right now working on this to fix the reviewd changes. Don't expect the optimal solution but I try to at least reach these goals:

- Be compatible with Flask < 2.2 where flask.json.JSONEncoder exists
- Be forward compatible with Flask >= 2.2 where the JSONEncoder generates deprecation warnings and be prepared once the decoder is removed.

I think this would be a good starting point to going further.

vimalloc commented:
> I'm right now working on this to fix the reviewd changes. Don't expect the optimal solution but I try to at least reach these goals:
> 
>     * Be compatible with Flask < 2.2 where flask.json.JSONEncoder exists
> 
>     * Be forward compatible with Flask >= 2.2 where the JSONEncoder generates deprecation warnings and be prepared once the decoder is removed.
> 
> 
> I think this would be a good starting point to going further.

Sounds great to me! I'll be around to help this get merged and a new release cut, and then digging into a more optimal solution can happen after. Thanks for your work on this!! 🥳 👍

jrast commented:
I pushed some new code:

- Works with flask <2.2 and flask >= 2.2
- Does currently not respect the `flask.json_provider_class` setting and uses conditional imports to get a valid encoder (see json_encoder.py)
- Coverage Test might fail as coverage is only run with one flask version and misses the conditional import code for the other flask version.

vimalloc commented on flask_jwt_extended/json_encoder.py:
```diff
@@ -0,0 +1,36 @@
+            return DefaultJSONProvider.default(o)
+
+except ModuleNotFoundError:  # pragma: no cover
+    from flask.json import JSONEncoder  # type: ignore  # noqa: F401
```
small thing, we should probably use `current_app.json_encoder` here so as not to break the current behavior on flask < 2.2

vimalloc commented on flask_jwt_extended/json_encoder.py:
```diff
@@ -0,0 +1,36 @@
+            overriding how basic types like ``str`` or ``list`` are
+            serialized, they are handled before this method.
+            """
+            return DefaultJSONProvider.default(o)
```
Looks like this line is the only thing missing from the coverage. Happy to have this be `# pragma: no cover` for now as well

jrast commented:
Damn, now you where faster than me... I was working on a simmilar idea, with the additional benefit that the new `app.json_provider_class` was respected in case of Flask >= 2.2....

vimalloc commented:
> Damn, now you where faster than me... I was working on a simmilar idea, with the additional benefit that the new `app.json_provider_class` was respected in case of Flask >= 2.2....

Sorry, didn't mean to step on your toes here, just had some time so figured I would take a stab at it! If you have a different idea please feel free to revert that last commit and push your stuff up!

jrast commented:
No problem at all! Nice to see other motivated devs working on this! Things in my latest commit:
- respect `app.json_provider_class`. However this requires some hacky way to generate a JSONEncoder class on the fly...
- moved from `json_encoder.py` to `internal_utils.py`
- Documented the stuff I did.

If this sound reasonable I will push the changes. I have also updated `tox.ini` to test with flask < 2.2 and flask > 2.2, I can also push this if you want.

vimalloc commented:
> No problem at all! Nice to see other motivated devs working on this! Things in my latest commit:
> 
>     * respect `app.json_provider_class`. However this requires some hacky way to generate a JSONEncoder class on the fly...
> 
>     * moved from `json_encoder.py` to `internal_utils.py`
> 
>     * Documented the stuff I did.
> 
> 
> If this sound reasonable I will push the changes. I have also updated `tox.ini` to test with flask < 2.2 and flask > 2.2, I can also push this if you want.

That would be awesome! Thank you for the very thorough work you're doing on this!! 👍

jrast commented:
Ok, I force-pushed my changes. Let's see what you think.

vimalloc commented on flask_jwt_extended/view_decorators.py:
```diff
@@ -86,8 +86,9 @@ def verify_jwt_in_request(
     # Should be impossible to hit, this makes mypy checks happy
-    if not _request_ctx_stack.top:  # pragma: no cover
-        raise RuntimeError("No _request_ctx_stack.top present, aborting")
+    # TODO: Convert this check to the new `flask.g` object
+    # if not _request_ctx_stack.top:  # pragma: no cover
+    #     raise RuntimeError("No _request_ctx_stack.top present, aborting")
```
This can be deleted now I think. Only reason this code existed was to satisfy mypy

vimalloc commented on flask_jwt_extended/internal_utils.py:
```diff
@@ -51,3 +64,31 @@ def custom_verification_for_token(jwt_header: dict, jwt_data: dict) -> None:
+    # use the method defined by the DefaultJSONProvider
+    default = getattr(app.json_provider_class, "default", DefaultJSONProvider.default)
+
+    class JSONEncoder(json.JSONEncoder):
+        def default(self, o: Any) -> Any:
+            return default(o)  # pragma: no cover
```
My only worry here is that I'm not sure if these ad-hoc created classes are going to get properly cleaned up in memory since one is basically being created every time we call `config.json_encoder`. I would hate to introduce a memory leak that caused problems in prod. Even though I agree that this looks prettier, I think maybe moving this back to the `try` statement at the top of the file would be safer. 

Thoughts?

jrast replied:
As we want to respect the `app.json_provider_class` I don't see a way to do this at import time as `current_app` is not valid then. I'm still thinking about a clean way, I'm not 100% happy with this either.

jrast replied:
Something like this:
````python
class JSONEncoder(json.JSONEncoder):
    def default(self, o: Any) -> Any:
        # If the registered JSON provider does not implement a default classmethod
        # use the method defined by the DefaultJSONProvider
        default = getattr(current_app.json_provider_class, "default", DefaultJSONProvider.default)
        return default(o)


def get_json_encoder(app: Flask) -> Type[json.JSONEncoder]:
    if not HAS_JSON_PROVIDER:  # pragma: no cover
        return app.json_encoder

    return JSONEncoder
````

vimalloc replied:
If it works, yeah!

jrast replied:
Seems to work. Added two tests (one for flask < 2.2, one for flask >= 2.2) to ensure this is working.

vimalloc commented:
Looks good. Left one small thought that I think is probably worth changing just to be safe, but overall I like the approach. Thanks!

Seluj78 reviewed (approved):
LGTM ! Can't wait to get this release :D

vimalloc reviewed (approved):
LGTM

Seluj78 commented:
Amazing ! Could you release this soon ? :)

vimalloc commented:
It's up as version `4.4.4`. Cheers!
